File: processor/src/main/java/com/michaelrocks/lightsaber/processor/files/ClassFileWriter.java
/*
 * Copyright 2015 Michael Rozumyanskiy
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.michaelrocks.lightsaber.processor.files;

import java.io.IOException;

public abstract class ClassFileWriter extends ClassFileVisitor implements AutoCloseable {
    public ClassFileWriter() {
        super(null);
    }

    @Override
    public void visitClassFile(final String path, final byte[] classData) throws IOException {
        writeFile(path, classData);
    }

    @Override
    public void visitOtherFile(final String path, final byte[] fileData) throws IOException {
        writeFile(path, fileData);
    }

    @Override
    public void visitDirectory(final String path) throws IOException {
        createDirectory(path);
    }

    protected abstract void writeFile(String path, byte[] fileData) throws IOException;

    protected abstract void createDirectory(String path) throws IOException;
}

File: processor/src/main/kotlin/io/michaelrocks/lightsaber/processor/io/ClassFileWriter.kt
/*
 * Copyright 2015 Michael Rozumyanskiy
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.michaelrocks.lightsaber.processor.io

import java.io.Closeable
import java.io.IOException

abstract class ClassFileWriter : ClassFileVisitor(null), Closeable {
  @Throws(IOException::class)
  override fun visitClassFile(path: String, classData: ByteArray) {
    writeFile(path, classData)
  }

  @Throws(IOException::class)
  override fun visitOtherFile(path: String, fileData: ByteArray) {
    writeFile(path, fileData)
  }

  @Throws(IOException::class)
  override fun visitDirectory(path: String) {
    createDirectory(path)
  }

  @Throws(IOException::class)
  protected abstract fun writeFile(path: String, fileData: ByteArray)

  @Throws(IOException::class)
  protected abstract fun createDirectory(path: String)
}
